File: src/lib/elaya/principal.ts
// Elaya principal resolver — identity → role + persona + permitted toolset.
//
// Every tool execution carries an ElayaPrincipal derived from the VERIFIED
// session profile (requireProfile / getCurrentProfile — A-01), never from
// model output or request payloads. Tools execute as this principal: identity
// args (userId/role/domain) passed to services are always principal-derived;
// the model only ever supplies filter values. Authorization is enforced in the
// tool layer + RLS — NEVER prompt-only.

import type { Profile, UserRole, AppDomain } from '@/lib/types';
import type { ElayaToolName } from '@/lib/elaya/tools/registry';
import { TOOLSET_BY_ROLE } from '@/lib/elaya/tools/registry';
import { CUSTOMER_TOOLSET, type ElayaCustomerToolName } from '@/lib/elaya/tools/customer-registry';
import type { GiaDomain } from '@/lib/constants/domains';

export type ElayaPersona = 'staff' | 'customer';

// ─────────────────────────────────────────────────────────────────────────
// ElayaPrincipal — a discriminated union. STAFF carries a verified profile +
// the role-gated read∪write toolset. CUSTOMER carries the LEAD identity (not a
// profile) + a HARD-CAPPED customer toolset (KB-send + interest-capture only,
// NO staff/CRM tools). The Golden Rule lives here in CODE: a customer's data
// scope + abilities come from this object, before the model runs — never from
// the prompt or any content the model reads (docs/architecture/elaya-jarvis-
// architecture.md, docs/modules/customer-welcome-blast.md).
// ─────────────────────────────────────────────────────────────────────────

export type StaffPrincipal = {
  kind: 'staff';
  persona: 'staff';
  userId: string;
  role: UserRole;
  domain: AppDomain;
  displayName: string;
  toolset: readonly ElayaToolName[];
};

export type CustomerPrincipal = {
  kind: 'customer';
  persona: 'customer';
  /** The lead row id — the ONLY record a customer turn may touch (its own lead). */
  leadId: string;
  /** The lead's Gia domain — scopes which training assets the customer toolset reads. */
  domain: GiaDomain;
  /** The customer's display name (first name), for a warm greeting. Never staff PII. */
  displayName: string;
  /** Hard-capped: send-company-material + capture-interest. NEVER a staff tool. */
  toolset: readonly ElayaCustomerToolName[];
};

export type ElayaPrincipal = StaffPrincipal | CustomerPrincipal;

/** Staff persona — verified profile → role + the role-gated toolset. */
export function resolveStaffPrincipal(profile: Profile): StaffPrincipal {
  return {
    kind: 'staff',
    persona: 'staff',
    userId: profile.id,
    role: profile.role,
    domain: profile.domain,
    displayName: profile.full_name,
    toolset: TOOLSET_BY_ROLE[profile.role],
  };
}

/**
 * Customer persona (FEATURE 2 — the WhatsApp customer channel). Identity is the
 * LEAD, never a profile. The toolset is the CUSTOMER_TOOLSET constant — a tiny,
 * fixed set that can NEVER read leads/deals/tasks/performance or any staff data
 * (those tools are not in it; the dispatch gate refuses anything outside it).
 * No prompt, no training content, no customer message can widen this — that is
 * the Golden Rule, enforced in code here.
 */
export function resolveCustomerPrincipal(lead: {
  id: string;
  domain: GiaDomain;
  first_name: string | null;
  last_name: string | null;
}): CustomerPrincipal {
  const displayName = (lead.first_name ?? '').trim() || 'there';
  return {
    kind: 'customer',
    persona: 'customer',
    leadId: lead.id,
    domain: lead.domain,
    displayName,
    toolset: CUSTOMER_TOOLSET,
  };
}

File: backend/app/brain/principal.py
"""The principal resolver — the Golden Rule in code (port of principal.ts).

Every tool execution carries a Principal derived from the VERIFIED profiles
row — never from model output, request payloads, or anything the model reads.
Tools execute AS this principal: identity args (user_id/role/domain) passed to
queries are always principal-derived; the model only ever supplies filter
values. Nothing a model reads can widen access.

Staff only for now — the customer persona ports with the WhatsApp flip.
"""

from __future__ import annotations

from dataclasses import dataclass

from app.core import supa
from app.tools.registry import TOOLSET_BY_ROLE

# The teams Elaya is switched on for (2026-09-26) — mirrors src/lib/constants/route-permissions.ts
# ELAYA_DOMAINS: the Gia sales domains and the concierge floor. Admin and founder always pass, and so
# does the tech workbench. A domain outside the list gets no principal, so no turn, whatever channel
# reached this brain (Node refuses at its doors first; this is the second lock).
ELAYA_DOMAINS = frozenset({"concierge", "onboarding", "house", "shop", "legacy"})
_WORKBENCH_DOMAINS = frozenset({"tech"})


def has_elaya_access(role: str, domain: str) -> bool:
    if role in ("admin", "founder"):
        return True
    if role == "guest":
        return False
    return domain in _WORKBENCH_DOMAINS or domain in ELAYA_DOMAINS


@dataclass(frozen=True)
class StaffPrincipal:
    user_id: str
    role: str
    domain: str
    display_name: str
    toolset: frozenset[str]
    # The concierge seat and its queendom (profiles.sia_role / queendom_id), for the reach hint only:
    # every bridged tool re-reads the queendom in Node at call time. None outside Concierge.
    sia_role: str | None = None
    queendom_id: str | None = None


async def resolve_staff_principal(user_id: str) -> StaffPrincipal | None:
    """Verified identity → role-gated toolset. Returns None for unknown or
    deactivated users — the caller refuses the turn entirely."""
    profile = await supa.get_profile(user_id)
    if profile is None or not profile.get("is_active", False):
        return None
    role = profile["role"]
    if not has_elaya_access(role, profile["domain"]):
        return None
    return StaffPrincipal(
        user_id=profile["id"],
        role=role,
        domain=profile["domain"],
        display_name=profile["full_name"],
        toolset=TOOLSET_BY_ROLE.get(role, frozenset()),
        sia_role=profile.get("sia_role"),
        queendom_id=profile.get("queendom_id"),
    )
